text/0680-implicit-injection-deprecation.md

```diff
@@ -0,0 +1,127 @@
+   However, because `store` is injected by default on all Ember.Route and Ember.Controller factory objects,
+   a user cannot easily opt out of `@ember-data/store`. This is a common problem space in other communities as well.
+   With previous JetBrains IDE installations, adding plugins required a reload. However, by removing a contructor function
+   that injected all the dependencies at once, they were able to avoid reload after installation of a plugin.
+
+2. Eager initialization of the dependency tree from which the property came from prevents tree shaking and incurs a
+   performance hit. For example, if you app does not need `ember-data` to render the entry route, the `store` injection will
+   take up more CPU cycles than necessary, hurting common user facing metrics.
+
+3. Eager initialization makes it hard to users to write tests that need to take advantage of a stub.
+   This is often a silent and annoying hurdle that users have to get around when writing tests.
+   Moreover, this is especially apparent for users moving to Ember's new testing APIs in [RFC 232](https://github.com/emberjs/rfcs/blob/master/text/0232-simplify-qunit-testing-api.md) and [RFC 268](https://github.com/emberjs/rfcs/blob/master/text/0268-acceptance-testing-refactor.md),
+   ensuring instance initializers run before each test.
+
+4. Injection on commonly used framework objects can lead to deviation of project specific styleguides. For instance,
+   injecting the `router` on the Ember.Component factory object will certainly lead to instances where the `router` is explicitly
+   injected and other instances where is it implicitly injected.
+
+5. Users have come to associate `Route.model()` with a hook that returns `ember-data` models in the absense of explicit injection.
+   While this can be true, it is not wholly true. New patterns of data loading are becoming
+   accepted in the community including opting to fetch data in a Component.
+
+
+## Detailed design
+
+Removing implicit injections is not possible until we first issue deprecations.
+This helps us maintain our strict semver commitment to not completing a major version bump and
+removing APIs without a deprecation path. As a result, it will take us 3 phases to remove `owner.inject`.
+
+### 1. Deprecate implicit injection on target object
+
+The first step will be issuing a deprecation on the target factory object that is receiving a property. This will surface the deprecation
+on user's owned objects including transitively through addons like `ember-data`. This can be accomplished by installing a native getter
+and setter on the target. Whenever the property is accessed, we will check if the property is explicitly injected. If it isn't,
+we will issue a deprecation with an until value, id and url to read more about this deprecation.
+
+### 2. Deprecate `owner.inject`
```

Comment:
Should mention how folks (like ember-data) should mitigate this deprecation. One possible answer is that ember-data can't remove its `owner.inject` until after ember-data@4. This ordering is a bit nebulous at the moment...

Reply:
FTR we eagerly inject only because of defaultStore. Were defaultStore not present we would deprecate this ourselves. Deprecating implicit injection without deprecating defaultStore is surprising to me, I'm hoping there was a follow up RFC merged that I've missed 😅

Reply:
I haven't seen `DS.defaultStore` anywhere for quite a while.  Does it still exist?

Reply:
@snewcomer it's not an EmberData concept, it's an Ember concept. You even linked my issue for needing the framework to remove it in the RFC but I think you possibly misunderstood it as something that we (EmberData) needed to remove. It is currently located here in Ember's codebase: https://github.com/emberjs/ember.js/blob/master/packages/%40ember/-internals/routing/lib/system/route.ts#L2323

Reply:
@runspired is I believe referring to this computed property: https://github.com/emberjs/ember.js/blob/master/packages/@ember/-internals/routing/lib/system/route.ts#L2319

We can and should also deprecate that property, though I think it will have to be for the 5.0 cycle. The deprecation warning should be enough to let user's know that they are using the deprecated functionality, and should explicitly inject `@store`, I think.

Reply:
@pzuraq the issue is that if we don't eagerly inject ourself the `find` call will not be the same as having called `ember-data's` `store.find` and results in either this assert triggering or (for folks that happen to have a `find` method on their model) a very confusing and unexpected request triggering.

```
assert(
            `${classify(name)} has no method \`find\`.`,
            typeof modelClass.find === 'function'
          );
```

Reply:
For folks that were previously making use of EmberData's injected store, they will get the deprecation and convert to using explicit service injection. For new apps in 4.0 and for places where devs simply forget to add in the explicit inject though the confusion will remain.

Reply:
Right, exactly. I don't think this will be a huge issue as long as we add a deprecation that says loudly in the console:

```
You are using the deprecated `store` computed property. This computed property is not a service, and it has confusing and usually unexpected behavior. If you are using Ember Data or another addon that provides a `store` service, you should add `@service('store')` to your route in order to get access to the store service instead.
```

I do agree ideally that the deprecation should be sooner rather than later ideally, but we'll have to see where it fits in the timeline.

Reply:
When deprecating default store we should deprecate implicit record loading in default `model` hook as well. There is a open RFC to do so #557.